### ISSUE TITLE
PrometheusCollector called for collection during initial creation

### ISSUE THREAD

**Issue by jamesachn** [open-telemetry/opentelemetry-java#4032]
**Describe the bug**
In 1.10.0 RC2, `SdkMeterProvider` appears to fail `.build()` if `PrometheusCollector` is a registered `MetricReader`.

The exact stack trace looks something like this:
```
Exception in thread "main" java.lang.IllegalStateException: No collection info for handle, this is a bug in the OpenTelemetry SDK.
	at io.opentelemetry.sdk.metrics.SdkMeterProvider$LeasedMetricProducer.collectAllMetrics(SdkMeterProvider.java:158)
	at io.opentelemetry.exporter.prometheus.PrometheusCollector.collect(PrometheusCollector.java:35)
	at io.prometheus.client.CollectorRegistry.collectorNames(CollectorRegistry.java:113)
	at io.prometheus.client.CollectorRegistry.register(CollectorRegistry.java:50)
	at io.prometheus.client.Collector.register(Collector.java:308)
	at io.prometheus.client.Collector.register(Collector.java:301)
	at io.opentelemetry.exporter.prometheus.PrometheusCollector$Factory.apply(PrometheusCollector.java:76)
	at io.opentelemetry.sdk.metrics.SdkMeterProvider.<init>(SdkMeterProvider.java:84)
	at io.opentelemetry.sdk.metrics.SdkMeterProviderBuilder.build(SdkMeterProviderBuilder.java:140)
```

What appears to be the case is that during `SdkMeterProvider` construction, all `MetricReaderFactory#apply()` are invoked for registered readers. For `PrometheusCollector.Factory`, as part of the call it will register itself w/ `CollectorRegistry.defaultRegistry`. But this registration prompts a callback to `PrometheusCollector#collect()` for metric data collection. This callback was previously "allowed" in prior release versions as the initial collection resolves to an empty collection. #3845 exposes and disallows this back-call to `SdkMeterProvider`, as technically, the `SdkMeterProvider` isn't fully initialized during this first back-call.

**Steps to reproduce**
```java
    public static void main(String[] args) {
        try {
            var provider = SdkMeterProvider.builder()
                    .registerMetricReader(PrometheusCollector.create())
                    .build();
            System.out.println("Provider created");
        } catch (Exception error) {
            error.printStackTrace();
        }
    }
```

**What did you expect to see?**
Snippet above should print to STDOUT `Provider created`.

**What did you see instead?**
Snippet prints the following to STDERR:
```
java.lang.IllegalStateException: No collection info for handle, this is a bug in the OpenTelemetry SDK.
	at io.opentelemetry.sdk.metrics.SdkMeterProvider$LeasedMetricProducer.collectAllMetrics(SdkMeterProvider.java:158)
	at io.opentelemetry.exporter.prometheus.PrometheusCollector.collect(PrometheusCollector.java:35)
	at io.prometheus.client.CollectorRegistry.collectorNames(CollectorRegistry.java:113)
	at io.prometheus.client.CollectorRegistry.register(CollectorRegistry.java:50)
	at io.prometheus.client.Collector.register(Collector.java:308)
	at io.prometheus.client.Collector.register(Collector.java:301)
	at io.opentelemetry.exporter.prometheus.PrometheusCollector$Factory.apply(PrometheusCollector.java:76)
	at io.opentelemetry.sdk.metrics.SdkMeterProvider.<init>(SdkMeterProvider.java:84)
	at io.opentelemetry.sdk.metrics.SdkMeterProviderBuilder.build(SdkMeterProviderBuilder.java:140)
	at example.Example.main(Example.java:13)
```

**What version and what artifacts are you using?**
Artifacts: `opentelemetry-api`, `opentelemetry-sdk`, `opentelemetry-exporter-prometheus`.
Version: `1.10.0`, `1.10.0-rc.2`
How did you reference these artifacts? `build.gradle`.

Snippets for `build.gradle`:
```gradle
dependencies {
    implementation(platform("io.opentelemetry:opentelemetry-bom:1.10.0-rc.2"))
    implementation("io.opentelemetry:opentelemetry-api")
    implementation("io.opentelemetry:opentelemetry-exporter-prometheus:1.10.0-alpha-rc.2")
    implementation("io.opentelemetry:opentelemetry-sdk")
    implementation("io.opentelemetry:opentelemetry-sdk-metrics:1.10.0-alpha-rc.2")

    implementation("io.prometheus:simpleclient_httpserver:0.12.0")
}
```
**Environment**
Compiler: openjdk version "17.0.1" (GraalVM CE 21.3.0 (build 17.0.1+12-jvmci-21.3-b05)
OS: MacOS 11.6.2

**Additional context**
Noticed this issue when bumping up dependency from `1.7.1` to `1.10.0-rc.2` to test out upcoming `1.10.0`.

**Comment by anuraaga** [comment]
@jamesachn Thanks for the report, this does seem to be a bug. At the day time we have been considering deprecating `PrometheusCollector` for removal - it depends on the unstable prometheus java library which is not a stable API. Are you able to use the newer `PrometheusHttpServer`?

**Comment by jamesachn** [comment]
I tested out w/ `PrometheusHttpServer`, and it doesn't exhibit the issue that `PrometheusCollector` had with this issue. Thank you!

One thing about `PrometheusHttpServer` though, is that (_I think_) in its current form it doesn't integrate as well with existing integration Prometheus's `CollectorRegistry`. Suppose that there are existing `Collector`s outside of application's integration with OpenTelemetry, with `PrometheusCollector` there is a piece-meal strategy of integrating with Open Telemetry metrics right now, while simultaneously working to migrate existing usage of Prometheus `Collector`s to Open Telemetry. But that strategy leverages the fact that both `PrometheusCollector` along with other "native" client `Collector`s currently in-use in application, by registering them all to a Prometheus `CollectorRegistry`, and then use Prometheus client lib's `HTTPServer` to serve up metrics to Prometheus scrapers.

That said, I think `PrometheusHttpServer` is a better and cleaner approach for my use-case as well, especially from resource-lifecycle control perspective as the piece-meal route is very messy to integrate into application's lifecycle. I was just curious to what, if any, other integration+migration approaches are available if `PrometheusCollector` is on its way to deprecation.

**Comment by anuraaga** [comment]
Thanks for digging @jamesachn - indeed integrating with other collection libraries is important to help with migrations.

Our general strategy is the opposite direction, have bridges from the other libraries into OTel. We are working on a micrometer bridge now and will definitely need to support other popular libraries like dropwizard. For our reference, do you use micrometer or the Prometheus client library directly? If that is the vast majority than we may not prioritize a bridge straight from the prom library but if it's commonly used we'll need it too at some point.

**Comment by jamesachn** [comment]
For the project in question that I am supporting, Prometheus client library is used directly.

That said, I think that choice was a matter of convenience and had some consideration of integration with existing infra in mind (already-made Grafana charts, for example). And I just did a quick survey on the project on the number of Prometheus client lib `Collector`s usage and outside of very few custom `Collector` which can be ported to interface with OTel directly, by in large usage of the standard "JVM metrics" collectors supported by Prometheus client library dominate the rest.

I noticed that OTel instrumentation library already implements metrics for GC and memory pool. It should be fairly simple to follow the same example and implement a port of Prometheus client's `DefaultExports`, leveraging `MBBean` in similar fashion, for my project's use-case.

So I think I'm good with `PrometheusHttpServer` going forward. _Most likely_, my project will use OTel directly as integration point. So I probably will not be utilizing the Micrometer bridge once all the migrations from direct Prometheus library completely transitions over.

Thank you for all the info! @anuraaga
